### PR TITLE
Added ModifyAttributeAction

### PR DESCRIPTION
**Pull request by grisstyl**
Nathan mentioned in a TODO for the Attributes wiki page that attribute actions were needed and I've come to want them in some spells I've been working on. This is my first undoing action and I modeled it after looking through some of the others so I may have done something wrong. Let me know!

### CODE REVIEW

**Comment by NathanWolf** [comment]
Thanks for doing this, I think it could be useful!

Looking it over, I kind of wonder if it should have some common base class with ModifyPropertiesAction? At the very least, it would be useful to have the equation-parsing elements in here.

File: Magic/pom.xml [hunk 1]
```diff
@@ -557,6 +557,14 @@
...
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
```
Comment (NathanWolf) [inline Magic/pom.xml]: We have discussed updating to Java 8, I probably will do it as part of the 1.13 compatibility release, but I think it should be a separate change, and not part of this PR.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ModifyAttributeAction.java [hunk 2]
```diff
@@ -48,6 +48,14 @@ public void prepare(CastContext context, ConfigurationSection parameters) {
         for (String key : section.getKeys(false)) {
             attributes.put(key, section.getDouble(key, 0));
         }
+
+        for (String key : parameters.getKeys(false)) {
+            if (key.startsWith("attribute_")) {
+                String str = key.replace("attribute_", "");
```
Comment (NathanWolf) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ModifyAttributeAction.java]: I have actually been trying to discourage this type of parametrization (such as in PotionEffectAction, I'd prefer add_effects versus effect_slow, etc). So I'm not sure about this part, sometimes fewer options can be better, or certainly less confusing.

File: Magic/pom.xml (hunk 1 above)
Comment (grisstyl) [reply Magic/pom.xml]: Odd, I thought I left that out of the PR. At one point the collection of the aggregate parameters used streams but I reverted that change. I guess IntelliJ doesn't revert the pom.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ModifyAttributeAction.java (hunk 2 above)
Comment (grisstyl) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/ModifyAttributeAction.java]: Will be removing this in the next commit.

**Comment by grisstyl** [comment]
Killme, what do you mean?

**Comment by NathanWolf** [comment]
That’s actually a general problem with several actions that I’m hoping to solve with the “status effect” system mentioned in the roadmap and todo.

I think in its current state this action handles it because it is only capable of decrementing and incrementing, but if we were to make it more generic it’d be a problem, assuming undo then has to become “set attribute back to its previous value”

The Fly action, ModifyWalkSpeed and ModifyProperties all have this problem. The status effect system is meant to solve this using a linked stack similar to the undo system.

**Comment by grisstyl** [comment]
I'd assume that we'd just be able to write the original contents to another Map unless you have a more efficient idea in mind. I like the idea of this status effects system that you mentioned under 7.7, but that release is reasonably far away considering you're planning on making 7.5 backwards compatible for 1.13.

**Comment by NathanWolf** [comment]
The status effect system is really only competing with new mob features at this point. The current dev build is about as 1.13-ready as I can make it without an actual server build to test against. (I may work on Alter and some other engineering magic issues, but those are pretty low priority)

That said, no it's not as simple as that, unfortunately.

The simplest example to consider is what if you have a spell that multiplies your strength by 150%, but it's duration is longer than its cooldown (bad design, but it's a simple example).

So you cast it once, let's say your strength is 10. It makes your strength 15, and stores 10 for undo.

Then you cast it again, it makes your strength 22.5 and remembers 15 for undo.

The first spell cast will undo first, it restores your strength to 10 but this effectively shortens the duration of the second cast.

Worse, though, when the second cast expires it will undo and restore your strength to 15, permanently.

The ModifyProperties action currently has this exact deficiency, which is why I only use it for changes that are meant to be permanent, such as buffing your wand. I would not recommend using it for temporary changes in its current state, though it does support undo because someone (akr on discord) specifically requested it even though I warned of the potential problems.

So I think each property (or more generically each status effect type in the status effect system- which could be properties, or attributes, or other random things like fly state, invulnerability, etc) needs to have a linked queue.

The undo system works like this, changes go onto a priority queue of sorts, so they can undo in order. But changes are also linked to one another in the order they were made, so each change knows to undo to the previous state. When a changes' undo fires it unlinks itself from its neighbors, so that we always preserve the initial state no matter the order, and changes know not to really apply their undo until they are the last change in the stack (to prevent early-undo of related changes).

It's complex to do properly, in other words, which is why the system has been on the TODO since like Magic 6 or so.

**Comment by grisstyl** [comment]
Ah, got it. I'm going to close this PR and move over to a branch as I should have done before.
